src/inferencer/inferencer.ts

```diff
@@ -286,27 +284,143 @@ export function inferProgram(program: es.Program): TypeAnnotatedNode<es.Program>
+    // e.g. T4 = T3
+    const block = functionDeclaration.body as TypeAnnotatedNode<es.BlockStatement>
+    const blockTypeVariable = block.typeVariable as Variable
+
+    if (blockTypeVariable !== undefined && prevReturnTypeVariable !== undefined) {
+      const result = updateTypeConstraints(blockTypeVariable, prevReturnTypeVariable)
+      if (result !== undefined) {
+        displayErrorAndTerminate(
+          'WARNING: There should not be a type error here in `inferFunctionDeclaration()` Part B - pls debug',
+          functionDeclaration.loc
+        )
+      }
+    }
+
+    // Finally, add constraint to give the function identifier the corresponding function type
+    // e.g. T5 = [T1] => T4
+    const iden = functionDeclaration.id as TypeAnnotatedNode<es.Identifier>
+    const idenTypeVariable = iden.typeVariable as Variable
+
+    const functionType = primitiveMap.get(iden.name) // Get function type from Type Env since it was added there
+
+    if (idenTypeVariable !== undefined && functionType !== undefined) {
+      const result = updateTypeConstraints(idenTypeVariable, functionType)
+      if (result !== undefined) {
+        displayErrorAndTerminate(
+          'WARNING: There should not be a type error here in `inferFunctionDeclaration()` Part C - pls debug',
+          functionDeclaration.loc
+        )
+      }
+    }
+  }
+
+  function inferFunctionApplication(functionApplication: TypeAnnotatedNode<es.CallExpression>) {
+    // Update type constraints in constraintStore
+    // e.g. Given: f^T5 (x^T1) { (return (...))^T2 ... (return (...))^T3 }^T4
+    //             f^T7 (1^T6)
+
+    // First, ensure arg nodes have same count as Γ(f)
+    // And try to add constraints that ensure arg nodes have same corresponding types
+    // e.g. T6 = T1
+    const iden = functionApplication.callee as TypeAnnotatedNode<es.Identifier>
+    const applicationArgs = functionApplication.arguments as TypeAnnotatedNode<es.Node>[]
+    const applicationArgCount = applicationArgs.length
+
+    const declarationFunctionType = primitiveMap.get(iden.name).types[0]
+    const declarationArgCount = declarationFunctionType.parameterTypes.length
+
+    if (applicationArgCount !== declarationArgCount) {
+      // check arg count
+      displayErrorAndTerminate(
+        `Expecting \`${declarationArgCount}\` arguments but got \`${applicationArgCount}\` instead`,
+        functionApplication.loc
+      )
+    }
+
+    for (let i = 0; i < applicationArgs.length; i++) {
+      // add type constraint for each arg
+      const applicationArgTypeVariable = applicationArgs[i].typeVariable as Variable
+      const declarationArgTypeVariable = declarationFunctionType.parameterTypes[i]
+        .typeVariable as Variable
+
+      if (applicationArgTypeVariable && declarationArgTypeVariable) {
+        const result = updateTypeConstraints(applicationArgTypeVariable, declarationArgTypeVariable)
+        if (result === -1) {
```

Comment:
This still checks for -1, when it should check for an error object

Reply:
Thanks for catching that, fixed